### PR TITLE
final edit bracket

### PR DESCRIPTION
**Pull request by wafiakmal**
Hi folks, in this pull request, I added the following:

1. `play_game()`, `give_character_choice`, `print_winner`.
2. Added some logic to determine the outcome of a duel (giving a print statement about the reason for winning if both output the same power).
3. New base variables `champion` and `total_rounds`.
4. Docstrings and minor mypy fixes.
5. Some prompts for playing the game on the terminal.
6. Minor .tsv data addition and typo fixing.

We can play the game on the terminal now. I've attached screenshots here.

image
image

### CODE REVIEW

File: src/game.py [hunk 1]
```diff
@@ -78,27 +93,119 @@ def duel(
                 + f"({chosen_spell['Description']})!"
             )
             print(
-                f"{character['Name']} has a total power rating of {battle_stats[i]}!\n"
+                f"{character['Name']} has a total power"
```
Comment (rjrichards27) [inline src/game.py:96]: Does this power score include the score form a spell?

File: src/game.py [hunk 2]
```diff
@@ -78,27 +93,119 @@ def duel(
...
+        if battle_stats[0] > battle_stats[1]:
+            winner_key = 0
+            print(
+                f"{character_group[winner_key]['Name']} defeated "
+                + f"{character_group[abs(winner_key-1)]['Name']}!\n\n"
+            )
+        elif battle_stats[0] == battle_stats[1]:
+            winner_key = 0
+            print(
+                f"{character_group[winner_key]['Name']} cast the spell "
```
Comment (rjrichards27) [inline src/game.py:111]: This is a cool way to handle ties

File: src/game.py [hunk 3]
```diff
@@ -111,22 +218,68 @@ def read_data(file_name: str) -> list[dict[str, str]]:
...
+        except ValueError:
+            print("Please enter a valid number!")
+
+    # returning the character choice
+    return characters[character_choice - 1]
+
+
 # reading in data for characters and spells
-hp_characters = "../data/Harry_Potter_Characters.tsv"
+hp_characters = (
```
Comment (rjrichards27) [inline src/game.py:249]: this path will only work for your computer so it should be the path to the data in the repo (as it was previously)

File: src/game.py (hunk 1 above)
Comment (wafiakmal) [reply src/game.py:96]: yes, because it's coming from `battle_stats[i]`

File: data/Harry_Potter_Characters.tsv [hunk 4]
```diff
@@ -52,6 +52,7 @@ Ginny Weasley,Daughter of Molly and Arthur.,5,5,7,6
...
 Percy Weasley,Third son of Arthur and Molly. He is a Gryffindor prefect.,5,4,8,4
 Ron Weasley,Harry's best friend.,6,6,6,7
-Rachel Richards, Hopeless Harry Potter Fan and Co-Creator of this Card Game,9,10,0,5
-Wafiakmal Mitah, Hopelessly Hopeless Harry Potter Fanboy Extraordinaire and Co-Creator of this Card Game,10,9,0,5
-Eric Rios, The One Guy Who Wanted this Card Game to be Anything but Harry Potter. Forced Co-Creator.,0,0,10,10
\ No newline at end of file
+Rachel Richards,Hopeless Harry Potter Fan and Co-Creator of this Card Game,9,10,0,5
+Wafiakmal Mitah,Hopelessly Hopeless Harry Potter Fanboy Extraordinaire and Co-Creator of this Card Game,10,9,0,5
+Eric Rios,The One Guy Who Wanted this Card Game to be Anything but Harry Potter. Forced Co-Creator.,0,0,10,10
+Patrick Wang,The newest professor of Defense Against the Dark Arts,7,8,10,5
```
Comment (EricR401S) [inline data/Harry_Potter_Characters.tsv:58]: This is beautiful, we can tell him to play the game until he finds that surprise!!!!!!!!!!!!!!!!!!!!!!!!!!!1

File: data/Harry_Potter_Spells.tsv [hunk 5]
```diff
@@ -1,28 +1,28 @@
 Name,Power,Description
 Avada Kedavra,1000000,Kills opponent
 Crucio,50,Causes extreme agony
-Imperio,50,Puts oppontent under caster's complete control
```
Comment (EricR401S) [inline data/Harry_Potter_Spells.tsv:4]: Yeah, this is good. Presentations is everything

Comment (EricR401S) [reply data/Harry_Potter_Spells.tsv:4]: Good fix

File: data/Harry_Potter_Spells.tsv [hunk 6]
```diff
@@ -1,28 +1,28 @@
...
+Langlock,3,Sticks opponent's tongue to the rook of their mouth
+Impediemta,5,Stops opponent in their tracks
 Salvio Hexia,7,Protects against hexes
-Stupefy,6,Stuns oppontent
-Silencio,5,Makes oppontent completely silent
+Stupefy,6,Stuns opponent
+Silencio,5,Makes opponent completely silent
 Protego,10,Protects caster
-Expelliarmus,9,Disarms oppontent
+Expelliarmus,9,Disarms opponent
```
Comment (EricR401S) [inline data/Harry_Potter_Spells.tsv:27]: Good work here, especially for the future print statements.

File: src/game.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+"""Provide user with a game of The Wizarding World of Harry Potter duels."""
```
Comment (EricR401S) [inline src/game.py:1]: Good catch.

File: src/game.py [hunk 8]
```diff
@@ -9,26 +10,36 @@ class Game:
     # intializing objects for class
     characters_playing: list[dict[str, str]]
     spells: list[dict[str, str]]
+    champion: str
+    total_rounds: int
 
-    def bracket_maker(self) -> list[dict[str, str]]:
-        """Creates bracket uses selected players"""
+    def bracket_maker(self) -> list[tuple[dict[str, str], dict[str, str]]]:
+        """Create bracket uses selected players."""
```
Comment (EricR401S) [inline src/game.py:17]: good catch here as well

File: src/game.py [hunk 9]
```diff
@@ -9,26 +10,36 @@ class Game:
...
         )
         return character_pairs
 
-    def print_duel_pairs(self, character_pairs_list: list[dict[str, str]]) -> None:
-        """Prints duel games before being played."""
+    def print_duel_pairs(
+        self, character_pairs_list: list[tuple[dict[str, str], dict[str, str]]]
+    ) -> None:
+        """Print duel games before being played."""
+        print("Behold of all the duels to come!\n")
```
Comment (EricR401S) [inline src/game.py:27]: nice touch

File: src/game.py [hunk 10]
```diff
@@ -62,7 +73,11 @@ def duel(
             if spell_num == len(spells_copy):
                 # this means the character used no spells
                 spell_stats = 0
-                chosen_spell = {"Name": "no spell", "Power": "0", "Description": ""}
+                chosen_spell = {
+                    "Name": "no spell",
+                    "Power": "0",
+                    "Description": "too scared to use a spell",
+                }
```
Comment (EricR401S) [inline src/game.py:80]: good format, don't know why my black didn't catch it.

File: src/game.py (hunk 2 above)
Comment (EricR401S) [reply src/game.py:111]: I implemented that in one of my drafts. I just wasn't sure to include it. I'm glad Wafi did.

File: src/game.py [hunk 11]
```diff
@@ -78,27 +93,119 @@ def duel(
...
+            print(
+                f"{character_group[winner_key]['Name']} defeated "
+                + f"{character_group[abs(winner_key-1)]['Name']}!\n\n"
+            )
+        elif battle_stats[0] == battle_stats[1]:
+            winner_key = 0
+            print(
+                f"{character_group[winner_key]['Name']} cast the spell "
+                + f"first and defeated "
+                + f"{character_group[abs(winner_key-1)]['Name']}!\n"
```
Comment (EricR401S) [inline src/game.py:113]: that was very clever, "first and defeated"

File: src/game.py [hunk 12]
```diff
@@ -78,27 +93,119 @@ def duel(
...
+    def play_game(self) -> None:
+        """Play the game."""
+        # creating the bracket
+        character_pairs_list = self.bracket_maker()
+        # playing the bracket
+        round = 1
+        temp_winners = []
+        while round <= self.total_rounds:
+            print(f"Prepare your wands! May the best Wizard/Witch win!\n")
+            # print("==================================================")
```
Comment (EricR401S) [inline src/game.py:143]: let me know if this should be erased later, this comment here

File: src/game.py [hunk 13]
```diff
@@ -78,27 +93,119 @@ def duel(
...
+                        f"break room, let's see the next set of duels:\n"
+                    )
+                    for duel_pairs in character_pairs_list:
+                        print(
+                            "Duel:",
+                            duel_pairs[0]["Name"],
+                            "vs.",
+                            duel_pairs[1]["Name"],
+                            "\n",
+                        )
```
Comment (EricR401S) [inline src/game.py:203]: very clean coding.

File: src/test_duel.py [hunk 14]
```diff
@@ -32,3 +32,5 @@ def test_duel() -> None:
     that are fated to battle each other."""
 
     assert type(game1.duel(battle_pair[0], test_spells)) is int
+    assert game1.duel(battle_pair[0], test_spells) in range(0, 3)
```
Comment (EricR401S) [inline src/test_duel.py:35]: Good tests.

File: src/test_game.py [hunk 15]
```diff
@@ -30,10 +31,44 @@ def test_reading_data() -> None:
...
+
+def test_play_game() -> None:
+    """Test the play game method."""
+    game1.play_game()
+    assert game1.characters_playing == 1
+
+
+def test_give_character_choice() -> None:
+    """Test the give character choice method."""
+    assert give_character_choice(selected_chars)
```
Comment (EricR401S) [inline src/test_game.py:74]: Great job on adding all the tests for the bracket methods.

**Review by EricR401S — APPROVED** [review]
I left many comments. Great job all across the board. We should show our tests to Patrick. @wafiakmal @rjrichards27  to make sure it meets the standards.

**Pull request merged**